### PR TITLE
Add error explanation for E0491

### PR DESCRIPTION
**Pull request by richardimaoka**
Refs #32777

Please let me know if any question or anything doesn't look right about this.

### CODE REVIEW

File: src/librustc/diagnostics.rs [hunk 1]
```diff
@@ -1454,6 +1454,42 @@ struct Prince<'kiss, 'SnowWhite: 'kiss> { // You say here that 'kiss must live
 ```
 "##,
 
+E0491: r##"
+A reference has a longer lifetime than the data it references.
+
+Erroneous code example:
+
+```compile_fail,E0491
+//struct containing a reference requires a lifetime parameter,
```
Comment (GuillaumeGomez) [inline src/librustc/diagnostics.rs]: Please add a whitespace after the '//'.

File: src/librustc/diagnostics.rs [hunk 2]
```diff
@@ -1454,6 +1454,42 @@ struct Prince<'kiss, 'SnowWhite: 'kiss> { // You say here that 'kiss must live
...
+//because the data the reference points to must outlive the struct (see E0106)
+struct Struct<'a> {
+    ref_i32: &'a i32
```
Comment (GuillaumeGomez) [inline src/librustc/diagnostics.rs]: Please add a comma at the end of the line.

File: src/librustc/diagnostics.rs [hunk 3]
```diff
@@ -1454,6 +1454,42 @@ struct Prince<'kiss, 'SnowWhite: 'kiss> { // You say here that 'kiss must live
...
+}
+
+//However, a nested struct like this, the signature itself does not tell
```
Comment (GuillaumeGomez) [inline src/librustc/diagnostics.rs]: Same.

File: src/librustc/diagnostics.rs [hunk 4]
```diff
@@ -1454,6 +1454,42 @@ struct Prince<'kiss, 'SnowWhite: 'kiss> { // You say here that 'kiss must live
...
+//whether 'a outlives 'b or the other way around.
+//So it could be possible that 'b of reference outlives 'a of the data.
+struct Nested<'a, 'b> {
+    ref_struct: &'b Struct<'a> //compile error E0491!
```
Comment (GuillaumeGomez) [inline src/librustc/diagnostics.rs]: Same (and same).

File: src/librustc/diagnostics.rs [hunk 5]
```diff
@@ -1454,6 +1454,42 @@ struct Prince<'kiss, 'SnowWhite: 'kiss> { // You say here that 'kiss must live
...
+}
+
+//'a: 'b means 'a outlives 'b
+struct Nested<'a: 'b, 'b> {
+    ref_struct: &'b Struct<'a>
+}
+```
+"##,
+
+
```
Comment (GuillaumeGomez) [inline src/librustc/diagnostics.rs]: Why so much empty lines?

File: src/librustc/diagnostics.rs [hunk 6]
```diff
@@ -1454,6 +1454,42 @@ struct Prince<'kiss, 'SnowWhite: 'kiss> { // You say here that 'kiss must live
...
+To fix this issue, you can specify a bound to the lifetime like below:
+
+```
+struct Struct<'a> {
+    ref_i32: &'a i32
+}
+
+//'a: 'b means 'a outlives 'b
+struct Nested<'a: 'b, 'b> {
+    ref_struct: &'b Struct<'a>
```
Comment (GuillaumeGomez) [inline src/librustc/diagnostics.rs]: Same.

File: src/librustc/diagnostics.rs [hunk 7]
```diff
@@ -1454,6 +1454,42 @@ struct Prince<'kiss, 'SnowWhite: 'kiss> { // You say here that 'kiss must live
...
+```
+
+To fix this issue, you can specify a bound to the lifetime like below:
+
+```
+struct Struct<'a> {
+    ref_i32: &'a i32
+}
+
+//'a: 'b means 'a outlives 'b
```
Comment (GuillaumeGomez) [inline src/librustc/diagnostics.rs]: Same.

**Review by GuillaumeGomez — CHANGES_REQUESTED** [review]
Excellent work! Just some little nits, nothing to say about the content itself.

File: src/librustc/diagnostics.rs [hunk 8]
```diff
@@ -1460,36 +1460,34 @@ A reference has a longer lifetime than the data it references.
...
 
-//However, a nested struct like this, the signature itself does not tell
-//whether 'a outlives 'b or the other way around.
-//So it could be possible that 'b of reference outlives 'a of the data.
+// However, a nested struct like this, the signature itself does not tell
+// whether 'a outlives 'b or the other way around.
+// So it could be possible that 'b of reference outlives 'a of the data.
 struct Nested<'a, 'b> {
-    ref_struct: &'b Struct<'a> //compile error E0491!
+    ref_struct: &'b Struct<'a>, //compile error E0491!
```
Comment (GuillaumeGomez) [inline src/librustc/diagnostics.rs]: Still missing a whitespace after '//'.

Comment (GuillaumeGomez) [inline src/librustc/diagnostics.rs]: Oh and the '!' isn't needed. ;)

**Review by GuillaumeGomez — CHANGES_REQUESTED** [review]
Just two nits remaining!

File: src/librustc/diagnostics.rs (hunk 8 above)
Comment (richard-imaoka) [inline src/librustc/diagnostics.rs]: ah sorry, fixing this now...

**Comment by richard-imaoka** [comment]
@GuillaumeGomez thank you very much for quick reviews. I think all your comments are reflected now.

**Comment by GuillaumeGomez** [comment]
Thanks! Now please [squash](https://blog.guillaume-gomez.fr/Git-tips/1/2) your commits and it'll be good to go!

**Review by GuillaumeGomez — APPROVED** [review]

File: src/librustc/diagnostics.rs [hunk 9]
```diff
@@ -1454,6 +1454,40 @@ struct Prince<'kiss, 'SnowWhite: 'kiss> { // You say here that 'kiss must live
...
+
+```
+struct Struct<'a> {
+    ref_i32: &'a i32,
+}
+
+// 'a: 'b means 'a outlives 'b
+struct Nested<'a: 'b, 'b> {
+    ref_struct: &'b Struct<'a>,
+}F
```
Comment (GuillaumeGomez) [inline src/librustc/diagnostics.rs]: Missed this one: Why is there an 'F' here?

**Review by GuillaumeGomez — CHANGES_REQUESTED** [review]

File: src/librustc/diagnostics.rs (hunk 9 above)
Comment (richard-imaoka) [inline src/librustc/diagnostics.rs]: oops sorry, :| will delete this F and squash my commits

**Comment by richard-imaoka** [comment]
Squashed into a single commit and removed the unnecessary 'F' as advised.

**Comment by GuillaumeGomez** [comment]
Thanks!

@bors: r+ rollup

**Comment by richard-imaoka** [comment]
Is this still in change requested state?

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Reviews status are just an indication, nothing to worry. I have put your PR in the bot's queue. It'll handle the tests and the merge for us.

**Pull request merged**